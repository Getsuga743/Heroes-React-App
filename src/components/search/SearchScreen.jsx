/* eslint-disable import/no-unresolved */
/* eslint-disable react/prop-types */
/* eslint-disable no-unused-vars */
/* eslint-disable react/jsx-props-no-spreading */
import React, { useMemo } from 'react';
import { useLocation } from 'react-router-dom';
import queryString from 'query-string';
import heroes from '../../data/heroes';
import useForm from '../../hooks/useForm';
import HeroCard from '../Heroes/HeroCard';
import getHeroesByName from '../../selectors/getHeroesByName';

const SearchScreen = ({ history }) => {
  const location = useLocation();
  // console.log(queryString.parse(location.search));
  const { q = '' } = queryString.parse(location.search);
  // extramos el nombre del heroe a buscar de la url, para eso usamos la librerÃ­a de queryString
  // eso es lo q le pasamos como argumento a la funcion getHeroesByName
  const [values, handleInputChange, reset] = useForm({
    searchText: q,
  });
  const { searchText } = values;

  const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);
  const handleSearch = (e) => {
    e.preventDefault();
    // console.log(searchText);
    // esto modifica la url de la pagina
    history.push(`?q=${searchText}`);
    reset();
  };
  return (
    <div>
      <h1>Search Screen</h1>
      <hr />
      <div className="row">
        <div className="col-5">
          <h4>Search</h4>
          <form onSubmit={handleSearch}>
            <input
              type="text"
              placeholder="Find your hero"
              className="form-control"
              onChange={handleInputChange}
              value={searchText}
              name="searchText"
              autoComplete="off"
            />
            <button
              type="submit"
              className="btn m-1 btn-block btn-outline-primary"
            >
              Search
            </button>
          </form>
        </div>
        <div className="col-7">
          <h4>Results</h4>
          <hr />
          {q === '' && (
            <div className="alert alert-info animate__animated animate__fadeIn">
              Search a hero
            </div>
          )}
          {q !== '' && heroesFiltered.length === 0 && (
            <div className="alert alert-danger  animate__animated animate__fadeIn">
              Hero not found...
            </div>
          )}

          {heroesFiltered.map((heroe) => (
            <HeroCard key={heroe.id} {...heroe} />
          ))}
        </div>
      </div>
    </div>
  );
};

export default SearchScreen;
