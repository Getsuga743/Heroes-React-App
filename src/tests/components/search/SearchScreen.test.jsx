import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import SearchScreen from '../../../components/search/SearchScreen';

describe('Pruebas en  <SearchScreen />', () => {
  test('debe de mostrarse correctamente con valores por defecto', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <Route path="/search" component={SearchScreen}></Route>
      </MemoryRouter>,
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.alert-info').text().trim()).toBe('Search a hero');
  });
  test('debe de mostrar a Batman y el input con el valor del queryString', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <Route path="/search" component={SearchScreen}></Route>
      </MemoryRouter>,
    );
    expect(wrapper.find('input').prop('value')).toBe('batman');
    expect(wrapper).toMatchSnapshot();
  });
  test('debe de mostrar un error si no se encuentra el heroe', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batmasdsad']}>
        <Route path="/search" component={SearchScreen}></Route>
      </MemoryRouter>,
    );
    expect(wrapper.find('.alert').exists()).toBe(true);
    expect(wrapper.find('.alert').text()).toBe('Hero not found...');
    expect(wrapper).toMatchSnapshot();
  });
  test('debe de llamar al push history', () => {
    const history = {
      push: jest.fn(),
    };
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <Route
          path="/search"
          component={() => <SearchScreen history={history} />}
        ></Route>
      </MemoryRouter>,
    );
    wrapper.find('input').simulate('change', {
      target: {
        name: 'searchText',
        value: 'batman',
      },
    });
    wrapper.find('form').simulate('submit', {
      preventDefault() {},
    });
    expect(history.push).toHaveBeenCalledWith('?q=batman');
  });
});
