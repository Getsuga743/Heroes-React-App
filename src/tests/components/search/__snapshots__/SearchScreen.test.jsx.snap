// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en  <SearchScreen /> debe de mostrar a Batman y el input con el valor del queryString 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="card ms-3  animate__animated animate__fadeIn"
        style={
          Object {
            "marginTop": "2rem",
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman portrait"
              className="card-img"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body align-items-end flex-column"
            >
              <h4
                className="card-title"
              >
                Batman
              </h4>
              <p
                className="card-text strong"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/heroe/dc-batman"
                onClick={[Function]}
              >
                Ver MÃ¡s
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en  <SearchScreen /> debe de mostrar un error si no se encuentra el heroe 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batmasdsad"
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger  animate__animated animate__fadeIn"
      >
        Hero not found...
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en  <SearchScreen /> debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info animate__animated animate__fadeIn"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
